### PR TITLE
Bitbucket Server: Add LoadPullRequest method to Client

### PR DESCRIPTION
**Pull request by mrnugget**
This adds support for Bitbucket Server pull requests to our Automation GraphQL API.

It allows users to use the `createChangeset` mutation with Bitbucket Server repositories and external IDs pointing to Bitbucket Server pull requests.

The user can then fetch the changesets via the GraphQL API and read every currently defined field for changesets pointing to Bitbucket Server pull requests.

It does that by adding
* a `LoadPullRequest` method to `bitbucketserver.Client`
* a `bitbucketserver.PullRequest` type definition
* a `LoadChangesets` method to the `BitbucketServerSource`
* the necessary `case` branches to read the necessary fields from a serialized Bitbucket Server pull request
* integration tests

### CODE REVIEW

File: enterprise/pkg/a8n/resolvers/graphql_test.go [hunk 1]
```diff
@@ -309,7 +330,16 @@ func TestCampaigns(t *testing.T) {
 		Changesets []Changeset
 	}
 
-	in := fmt.Sprintf(`[{repository: %q, externalID: %q}]`, graphqlRepoID, "999")
+	graphqlGithubRepoID := string(marshalRepositoryID(api.RepoID(githubRepo.ID)))
+	graphqlBBSRepoID := string(marshalRepositoryID(api.RepoID(bbsRepo.ID)))
+	fmt.Println(graphqlBBSRepoID)
```
Comment (tsenart) [inline enterprise/pkg/a8n/resolvers/graphql_test.go]:
```suggestion
```

File: pkg/extsvc/bitbucketserver/client.go [hunk 2]
```diff
@@ -714,6 +729,50 @@ type Project struct {
 	} `json:"links"`
 }
 
+type Ref struct {
+	ID         string `json:"id"`
+	Repository *Repo  `json:"repository"`
```
Comment (tsenart) [inline pkg/extsvc/bitbucketserver/client.go]: From the documentation, here's the JSON that's returned from the API. By using a full repository representation here, we'd be wasting storage in the changesets table I believe. Is there a benefit in doing so?

```json
{
    "id": 101,
    "version": 1,
    "title": "Talking Nerdy",
    "description": "It’s a kludge, but put the tuple from the database in the cache.",
    "state": "OPEN",
    "open": true,
    "closed": false,
    "createdDate": 1359075920,
    "updatedDate": 1359085920,
    "fromRef": {
        "id": "refs/heads/feature-ABC-123",
        "repository": {
            "slug": "my-repo",
            "name": null,
            "project": {
                "key": "PRJ"
            }
        }
    },
    "toRef": {
        "id": "refs/heads/master",
        "repository": {
            "slug": "my-repo",
            "name": null,
            "project": {
                "key": "PRJ"
            }
        }
    },
    "locked": false,
    "author": {
        "user": {
            "name": "tom",
            "emailAddress": "tom@example.com",
            "id": 115026,
            "displayName": "Tom",
            "active": true,
            "slug": "tom",
            "type": "NORMAL"
        },
        "role": "AUTHOR",
        "approved": true,
        "status": "APPROVED"
    },
    "reviewers": [
        {
            "user": {
                "name": "jcitizen",
                "emailAddress": "jane@example.com",
                "id": 101,
                "displayName": "Jane Citizen",
                "active": true,
                "slug": "jcitizen",
                "type": "NORMAL"
            },
            "lastReviewedCommit": "7549846524f8aed2bd1c0249993ae1bf9d3c9998",
            "role": "REVIEWER",
            "approved": true,
            "status": "APPROVED"
        }
    ],
    "participants": [
        {
            "user": {
                "name": "dick",
                "emailAddress": "dick@example.com",
                "id": 3083181,
                "displayName": "Dick",
                "active": true,
                "slug": "dick",
                "type": "NORMAL"
            },
            "role": "PARTICIPANT",
            "approved": false,
            "status": "UNAPPROVED"
        },
        {
            "user": {
                "name": "harry",
                "emailAddress": "harry@example.com",
                "id": 99049120,
                "displayName": "Harry",
                "active": true,
                "slug": "harry",
                "type": "NORMAL"
            },
            "role": "PARTICIPANT",
            "approved": true,
            "status": "APPROVED"
        }
    ],
    "links": {
        "self": [
            {
                "href": "http://link/to/pullrequest"
            }
        ]
    }
```

Comment (mrnugget) [reply pkg/extsvc/bitbucketserver/client.go]: What do you think we should get rid off? And, what if we decide in the future that we need it? Is it worth saving on storage now?

File: cmd/repo-updater/repos/bitbucketserver_test.go [hunk 3]
```diff
@@ -194,6 +199,108 @@ func TestBitbucketServerSource_Exclude(t *testing.T) {
...
+			lg.SetHandler(log15.DiscardHandler())
+
+			svc := &ExternalService{
+				Kind: "BITBUCKETSERVER",
+				Config: marshalJSON(t, &schema.BitbucketServerConnection{
+					// The test fixtures and golden files were generated with
+					// this config pointed to bitbucket.sgdev.org
+					Url:             os.Getenv("BITBUCKET_SERVER_URL"),
+					Token:           os.Getenv("BITBUCKET_SERVER_TOKEN"),
+					RepositoryQuery: []string{"all"},
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply cmd/repo-updater/repos/bitbucketserver_test.go]: Nope! Nice catch!

**Comment by mrnugget** [comment]
> What's the next step on Bitbucket Server support? Webhooks?

That depends. Strictly speaking we don't "need" the webhooks right now.

**Pull request merged**